src/utilities/introspectionQuery.js

```diff
@@ -124,95 +124,117 @@ export type IntrospectionType =
   | IntrospectionEnumType
   | IntrospectionInputObjectType;
 
-export type IntrospectionScalarType = {
-  kind: 'SCALAR';
-  name: string;
-  description: ?string;
-};
-
-export type IntrospectionObjectType = {
-  kind: 'OBJECT';
-  name: string;
-  description: ?string;
-  fields: Array<IntrospectionField>;
-  interfaces: Array<IntrospectionNamedTypeRef>;
-};
-
-export type IntrospectionInterfaceType = {
-  kind: 'INTERFACE';
-  name: string;
-  description: ?string;
-  fields: Array<IntrospectionField>;
-  possibleTypes: Array<IntrospectionNamedTypeRef>;
-};
-
-export type IntrospectionUnionType = {
-  kind: 'UNION';
-  name: string;
-  description: ?string;
-  possibleTypes: Array<IntrospectionNamedTypeRef>;
-};
-
-export type IntrospectionEnumType = {
-  kind: 'ENUM';
-  name: string;
-  description: ?string;
-  enumValues: Array<IntrospectionEnumValue>;
-};
-
-export type IntrospectionInputObjectType = {
-  kind: 'INPUT_OBJECT';
-  name: string;
-  description: ?string;
-  inputFields: Array<IntrospectionInputValue>;
-};
+export type IntrospectionOutputType =
+  | IntrospectionScalarType
+  | IntrospectionObjectType
+  | IntrospectionInterfaceType
+  | IntrospectionUnionType
+  | IntrospectionEnumType;
+
+export type IntrospectionInputType =
+  | IntrospectionScalarType
+  | IntrospectionEnumType
+  | IntrospectionInputObjectType;
+
+export type IntrospectionScalarType = {|
```

Comment:
I'm not sure if these types should be exact types since that doesn't actually describe what is returned from the server - all fields are queried on each possible type, so the properties exist and are null.

That might be okay as just a balance between correctness and usability, but we should keep an eye on this in case we need to remove the exact types in the future.

Reply:
I'd prefer exact types for now, as it's way easier to pretend fields that are always null don't exist. If we have to add a bunch of null fields, we might want a conversion from `IntrospectionType` to `LinkedType` to give us the more-restrictive typing (could also do things like enforce `ofType` is non-null on `List` and `NonNullable`).

Basically, if we need to relax this, then we probably want two different sets of types:

`IntrospectionType` => things coming over-the-wire in an introspection query. Does not differentiate from `Input` and `Output` types, but is a 1:1 data representation.

`LinkedType` => type system description using reference types that is a validated version of what came down from the wire.

Reply:
I think it's fine to keep it exact types for now - just wanted to highlight the disconnect so we're aware in the future